File: pdu/python/hako_msgs/pdu_conv_SimpleStructVarray.py

import struct
from .pdu_pytype_SimpleStructVarray import SimpleStructVarray
from ..pdu_utils import *
from .. import binary_io

# dependencies for the generated Python class
from ..hako_msgs.pdu_conv_SimpleVarray import *
from ..hako_msgs.pdu_conv_SimpleVarray import *



def pdu_to_py_SimpleStructVarray(binary_data: bytearray) -> SimpleStructVarray:
    py_obj = SimpleStructVarray()
    meta_parser = binary_io.PduMetaDataParser()
    meta = meta_parser.load_pdu_meta(binary_data)
    if meta is None:
        raise ValueError("Invalid PDU binary data: MetaData not found or corrupted")
    binary_read_recursive_SimpleStructVarray(meta, binary_data, py_obj, binary_io.PduMetaData.PDU_META_DATA_SIZE)
    return py_obj


def binary_read_recursive_SimpleStructVarray(meta: binary_io.PduMetaData, binary_data: bytearray, py_obj: SimpleStructVarray, base_off: int):
    # array_type: single 
    # data_type: primitive 
    # member_name: aaa 
    # type_name: int32 
    # offset: 0 size: 4 
    # array_len: 1

    
    bin = binary_io.readBinary(binary_data, base_off + 0, 4)
    py_obj.aaa = binary_io.binTovalue("int32", bin)
    
    # array_type: array 
    # data_type: primitive 
    # member_name: fixed_str 
    # type_name: string 
    # offset: 4 size: 256 
    # array_len: 2

    
    array_value = binary_io.readBinary(binary_data, base_off + 4, 256)
    py_obj.fixed_str = binary_io.binToArrayValues("string", array_value)
    
    # array_type: varray 
    # data_type: primitive 
    # member_name: varray_str 
    # type_name: string 
    # offset: 260 size: 128 
    # array_len: 8

    array_size = binary_io.binTovalue("int32", binary_io.readBinary(binary_data, base_off + 260, 4))
    offset_from_heap = binary_io.binTovalue("int32", binary_io.readBinary(binary_data, base_off + 260 + 4, 4))
    one_elm_size = 128 
    array_value = binary_io.readBinary(binary_data, meta.heap_off + offset_from_heap, one_elm_size * array_size)
    py_obj.varray_str = array_value
    
    # array_type: array 
    # data_type: struct 
    # member_name: fixed_array 
    # type_name: SimpleVarray 
    # offset: 268 size: 120 
    # array_len: 5

    i = 0
    array_size = 5
    one_elm_size = int(120 / array_size)
    array_value = []
    while i < array_size:
        tmp_py_obj = SimpleVarray()
        binary_read_recursive_SimpleVarray(meta, binary_data, tmp_py_obj, base_off + 268 + (i * one_elm_size))
        array_value.append(tmp_py_obj)
        i = i + 1
    py_obj.fixed_array = array_value    
    
    # array_type: varray 
    # data_type: struct 
    # member_name: data 
    # type_name: SimpleVarray 
    # offset: 388 size: 24 
    # array_len: 8

    array_size = binary_io.binTovalue("int32", binary_io.readBinary(binary_data, base_off + 388, 4))
    offset_from_heap = binary_io.binTovalue("int32", binary_io.readBinary(binary_data, base_off + 388 + 4, 4))
    one_elm_size = 24
    i = 0
    array_value = []
    while i < array_size:
        tmp_py_obj = SimpleVarray()
        binary_read_recursive_SimpleVarray(meta, binary_data, tmp_py_obj, meta.heap_off + offset_from_heap + (i * one_elm_size))
        array_value.append(tmp_py_obj)
        i = i + 1
    py_obj.data = array_value    
    
    return py_obj


def py_to_pdu_SimpleStructVarray(py_obj: SimpleStructVarray) -> bytearray:
    binary_data = bytearray()
    base_allocator = DynamicAllocator(False)
    bw_container = BinaryWriterContainer(binary_io.PduMetaData())
    binary_write_recursive_SimpleStructVarray(0, bw_container, base_allocator, py_obj)

    # メタデータの設定
    total_size = base_allocator.size() + bw_container.heap_allocator.size() + binary_io.PduMetaData.PDU_META_DATA_SIZE
    bw_container.meta.total_size = total_size
    bw_container.meta.heap_off = binary_io.PduMetaData.PDU_META_DATA_SIZE + base_allocator.size()

    # binary_data のサイズを total_size に調整
    if len(binary_data) < total_size:
        binary_data.extend(bytearray(total_size - len(binary_data)))
    elif len(binary_data) > total_size:
        del binary_data[total_size:]

    # メタデータをバッファにコピー
    binary_io.writeBinary(binary_data, 0, bw_container.meta.to_bytes())

    # 基本データをバッファにコピー
    binary_io.writeBinary(binary_data, bw_container.meta.base_off, base_allocator.to_array())

    # ヒープデータをバッファにコピー
    binary_io.writeBinary(binary_data, bw_container.meta.heap_off, bw_container.heap_allocator.to_array())

    return binary_data

def binary_write_recursive_SimpleStructVarray(parent_off: int, bw_container: BinaryWriterContainer, allocator, py_obj: SimpleStructVarray):
    # array_type: single 
    # data_type: primitive 
    # member_name: aaa 
    # type_name: int32 
    # offset: 0 size: 4 
    # array_len: 1
    type = "int32"
    off = 0

    
    bin = binary_io.typeTobin(type, py_obj.aaa)
    bin = get_binary(type, bin, 4)
    allocator.add(bin, expected_offset=parent_off + off)
    
    # array_type: array 
    # data_type: primitive 
    # member_name: fixed_str 
    # type_name: string 
    # offset: 4 size: 256 
    # array_len: 2
    type = "string"
    off = 4

    
    elm_size =  256 
    array_size = int(128.0)
    one_elm_size = int(elm_size / array_size)
    binary = binary_io.typeTobin_array(type, py_obj.fixed_str, one_elm_size)
    allocator.add(binary, expected_offset=(parent_off + off))
    
    # array_type: varray 
    # data_type: primitive 
    # member_name: varray_str 
    # type_name: string 
    # offset: 260 size: 128 
    # array_len: 8
    type = "string"
    off = 260

    offset_from_heap = bw_container.heap_allocator.size()
    array_size = len(py_obj.varray_str)
    binary = binary_io.typeTobin_array(type, py_obj.varray_str, 128)
    bw_container.heap_allocator.add(binary, expected_offset=0)
    a_b = array_size.to_bytes(4, byteorder='little')
    o_b = offset_from_heap.to_bytes(4, byteorder='little')
    allocator.add(a_b + o_b, expected_offset=parent_off + off)
    
    # array_type: array 
    # data_type: struct 
    # member_name: fixed_array 
    # type_name: SimpleVarray 
    # offset: 268 size: 120 
    # array_len: 5
    type = "SimpleVarray"
    off = 268

    for i, elm in enumerate(py_obj.fixed_array):
        elm_size = 120
        array_size = int(24.0)
        one_elm_size = int(elm_size / array_size)
        binary_write_recursive_SimpleVarray((parent_off + off + i * one_elm_size), bw_container, allocator, elm)
    
    # array_type: varray 
    # data_type: struct 
    # member_name: data 
    # type_name: SimpleVarray 
    # offset: 388 size: 24 
    # array_len: 8
    type = "SimpleVarray"
    off = 388

    offset_from_heap = bw_container.heap_allocator.size()
    array_size = len(py_obj.data)
    for i, elm in enumerate(py_obj.data):
        one_elm_size =  24
        binary_write_recursive_SimpleVarray((parent_off + i * one_elm_size), bw_container, bw_container.heap_allocator, elm)
    a_b = array_size.to_bytes(4, byteorder='little')
    o_b = offset_from_heap.to_bytes(4, byteorder='little')
    allocator.add(a_b + o_b, expected_offset=parent_off + off)    
    

if __name__ == "__main__":
    import sys
    import json

    def print_usage():
        print(f"Usage: python -m pdu.python.pdu_conv_SimpleStructVarray <read|write> [args...]")
        print(f"  read <input_binary_file> <output_json_file>")
        print(f"  write <input_json_file> <output_binary_file>")

    if len(sys.argv) < 2:
        print_usage()
        sys.exit(1)

    command = sys.argv[1]

    if command == "read":
        if len(sys.argv) != 4:
            print_usage()
            sys.exit(1)
        
        binary_filepath = sys.argv[2]
        output_json_filepath = sys.argv[3]

        with open(binary_filepath, "rb") as f:
            binary_data = bytearray(f.read())
        
        py_obj = pdu_to_py_SimpleStructVarray(binary_data)
        
        with open(output_json_filepath, "w") as f:
            f.write(py_obj.to_json())

    elif command == "write":
        if len(sys.argv) != 4:
            print_usage()
            sys.exit(1)

        input_json_filepath = sys.argv[2]
        output_binary_filepath = sys.argv[3]

        with open(input_json_filepath, "r") as f:
            json_str = f.read()
        
        py_obj = SimpleStructVarray.from_json(json_str)
        
        binary_data = py_to_pdu_SimpleStructVarray(py_obj)

        with open(output_binary_filepath, "wb") as f:
            f.write(binary_data)

    else:
        print(f"Unknown command: {command}")
        print_usage()
        sys.exit(1)

File: pdu/javascript/hako_msgs/pdu_conv_SimpleStructVarray.js
import * as PduUtils from '../pdu_utils.js';
import { SimpleStructVarray } from './pdu_jstype_SimpleStructVarray.js';
import { SimpleVarray } from './pdu_jstype_SimpleVarray.js';
import { binary_read_recursive_SimpleVarray, binary_write_recursive_SimpleVarray } from './pdu_conv_SimpleVarray.js';


/**
 * Deserializes a binary PDU into a SimpleStructVarray object.
 * @param {ArrayBuffer} binary_data
 * @returns { SimpleStructVarray }
 */
export function pduToJs_SimpleStructVarray(binary_data) {
    const js_obj = new SimpleStructVarray();
    const meta_parser = new PduUtils.PduMetaDataParser();
    const meta = meta_parser.load_pdu_meta(binary_data);
    if (meta === null) {
        throw new Error("Invalid PDU binary data: MetaData not found or corrupted");
    }
    binary_read_recursive_SimpleStructVarray(meta, binary_data, js_obj, meta.base_off);
    return js_obj;
}

export function binary_read_recursive_SimpleStructVarray(meta, binary_data, js_obj, base_off) {
    const view = new DataView(binary_data);
    const littleEndian = true;
    // member: aaa, type: int32 (primitive)

    
    {
        const bin = PduUtils.readBinary(binary_data, base_off + 0, 4);
        js_obj.aaa = PduUtils.binToValue("int32", bin);
    }
    
    // member: fixed_str, type: string (primitive)

    
    {
        const array_bin = PduUtils.readBinary(binary_data, base_off + 4, 256);
        js_obj.fixed_str = PduUtils.binToArrayValues("string", array_bin, 2);
    }
    
    // member: varray_str, type: string (primitive)

    { // varray
        const array_size = view.getInt32(base_off + 260, littleEndian);
        const offset_from_heap = view.getInt32(base_off + 260 + 4, littleEndian);
        const one_elm_size = 128;
        const array_bin = PduUtils.readBinary(binary_data, meta.heap_off + offset_from_heap, one_elm_size * array_size);
        
        if ("string" === 'string') {
            js_obj.varray_str = PduUtils.binToValue("string", array_bin);
        } else {
            js_obj.varray_str = PduUtils.binToArrayValues("string", array_bin, array_size);
        }
    }
    
    // member: fixed_array, type: SimpleVarray (struct)

    {
        const array_size = 5;
        const one_elm_size = 120 / array_size;
        const array_value = [];
        for (let i = 0; i < array_size; i++) {
            const tmp_obj = new SimpleVarray();
            const item_offset = base_off + 268 + (i * one_elm_size);
            binary_read_recursive_SimpleVarray(meta, binary_data, tmp_obj, item_offset);
            array_value.push(tmp_obj);
        }
        js_obj.fixed_array = array_value;
    }
    
    // member: data, type: SimpleVarray (struct)

    { // varray
        const array_size = view.getInt32(base_off + 388, littleEndian);
        const offset_from_heap = view.getInt32(base_off + 388 + 4, littleEndian);
        const one_elm_size = 24;
        const array_value = [];
        for (let i = 0; i < array_size; i++) {
            const tmp_obj = new SimpleVarray();
            const item_offset = meta.heap_off + offset_from_heap + (i * one_elm_size);
            binary_read_recursive_SimpleVarray(meta, binary_data, tmp_obj, item_offset);
            array_value.push(tmp_obj);
        }
        js_obj.data = array_value;
    }
    
    return js_obj;
}

/**
 * Serializes a SimpleStructVarray object into a binary PDU.
 * @param { SimpleStructVarray } js_obj
 * @returns {ArrayBuffer}
 */
export function jsToPdu_SimpleStructVarray(js_obj) {
    const base_allocator = new PduUtils.DynamicAllocator();
    const bw_container = new PduUtils.BinaryWriterContainer(new PduUtils.PduMetaData());

    binary_write_recursive_SimpleStructVarray(0, bw_container, base_allocator, js_obj);

    const base_data_size = base_allocator.size();
    const heap_data_size = bw_container.heap_allocator.size();
    
    bw_container.meta.heap_off = PduUtils.PDU_META_DATA_SIZE + base_data_size;
    bw_container.meta.total_size = bw_container.meta.heap_off + heap_data_size;

    const final_buffer = new ArrayBuffer(bw_container.meta.total_size);
    const final_view = new Uint8Array(final_buffer);

    PduUtils.writeBinary(final_view, 0, bw_container.meta.to_bytes());
    PduUtils.writeBinary(final_view, bw_container.meta.base_off, base_allocator.toArray());
    PduUtils.writeBinary(final_view, bw_container.meta.heap_off, bw_container.heap_allocator.toArray());

    return final_buffer;
}

export function binary_write_recursive_SimpleStructVarray(parent_off, bw_container, allocator, js_obj) {
    const littleEndian = true;
    // member: aaa, type: int32 (primitive)

    
    {
        const bin = PduUtils.typeToBin("int32", js_obj.aaa, 4);
        allocator.add(bin, parent_off + 0);
    }
    
    // member: fixed_str, type: string (primitive)

    
    {
        const buffer = PduUtils.typesToBin("string", js_obj.fixed_str);
        allocator.add(buffer, parent_off + 4);
    }
    
    // member: varray_str, type: string (primitive)

    { // varray
        const offset_from_heap = bw_container.heap_allocator.size();
        let data_buffer;
        let array_size;
        if ("string" === 'string') {
            data_buffer = new TextEncoder().encode(js_obj.varray_str);
            array_size = data_buffer.byteLength;
        } else {
            data_buffer = PduUtils.typesToBin("string", js_obj.varray_str);
            array_size = js_obj.varray_str.length;
        }
        bw_container.heap_allocator.add(data_buffer);

        const ref_buffer = new ArrayBuffer(8);
        const ref_view = new DataView(ref_buffer);
        ref_view.setInt32(0, array_size, littleEndian);
        ref_view.setInt32(4, offset_from_heap, littleEndian);
        allocator.add(ref_buffer, parent_off + 260);
    }
    
    // member: fixed_array, type: SimpleVarray (struct)

    {
        const one_elm_size = 120 / 5;
        for (let i = 0; i < 5; i++) {
            binary_write_recursive_SimpleVarray(parent_off + 268 + (i * one_elm_size), bw_container, allocator, js_obj.fixed_array[i]);
        }
    }
    
    // member: data, type: SimpleVarray (struct)

    { // varray
        const offset_from_heap = bw_container.heap_allocator.size();
        const one_elm_size = 24;
        for (const elm of js_obj.data) {
            binary_write_recursive_SimpleVarray(0, bw_container, bw_container.heap_allocator, elm);
        }

        const ref_buffer = new ArrayBuffer(8);
        const ref_view = new DataView(ref_buffer);
        ref_view.setInt32(0, js_obj.data.length, littleEndian); // array_size
        ref_view.setInt32(4, offset_from_heap, littleEndian);
        allocator.add(ref_buffer, parent_off + 388);
    }
    
}
